modules/saku-policy-store/src/saku/schemas.clj

```diff
@@ -1,53 +1,198 @@
 (ns saku.schemas
-  (:require [malli.core :as m]))
+  "Malli schemas for all Saku API operations."
+  (:require
+    [camel-snake-kebab.core :as csk]
+    [clojure.string :as str]
+    [malli.core :as m]
+    [malli.transform :as mt]
+    [malli.util :as mu]))
 
-(def effect
-  (m/schema [:tuple [:= :effect] [:enum :deny :allow]]))
+(defn rename-keys
+  [m kmap]
+  (reduce-kv
+    (fn [m' k-src fn-or-k-target]
+      (let [v-src (get m k-src)
+            [k-target v] (if (fn? fn-or-k-target)
+                           (fn-or-k-target v-src)
+                           [fn-or-k-target v-src])]
+        (-> m'
+          (cond-> v (assoc k-target v))
+          (cond-> (not= k-src k-target) (dissoc k-src)))))
+    m kmap))
 
-(def identity-statement
-  (m/schema [:map
-             [:statement/actions [:sequential :string]]
-             [:statement/effect effect]
-             [:statement/resources [:sequential :string]]]))
-
-(def resource-statement
-  (m/schema [:map
-             [:statement/actions [:sequential :string]]
-             [:statement/effect effect]
-             [:statement/identities [:sequential :string]]]))
+(defn- -compile-rename-keys-transformer
+  [kmap]
+  {:leave (if (seq kmap)
+            #(rename-keys % kmap)
+            identity)})
 
-(def resource-policy
-  (m/schema [:map
-             [:policy/drn :string]
-             [:policy/statements [:sequential {:min 1} resource-statement]]]))
-
-(def resource-policies
-  (m/schema [:sequential resource-policy]))
-
-(def identity-policy
-  (m/schema [:map
-             [:policy/drn :string]
-             [:policy/statements [:sequential {:min 1} identity-statement]]]))
+(defn rename-keys-transformer
+  [{:keys [rename-key]}]
+  (mt/transformer
+    {:decoders
+     {:map
+      {:compile
+       (fn [schema _]
+         (-compile-rename-keys-transformer
+           (into {}
+             (keep (fn [[k-src val-schema]]
+                     (when-let [rename (get (m/properties val-schema) rename-key)]
+                       [k-src rename])))
+             (m/entries schema))))}}}))
```

Comment:
The "guts" of the transformer. Really quite simple. I actually think there's a general data transformation engine hidden behind the scenes here, but it'd take a bit of finessing to figure out.